### PR TITLE
US040 - Login System

### PR DESCRIPTION
**Pull request by joel-and**

### CODE REVIEW

**Review by BenjaminBrashaw — APPROVED** [review]

**Review by BenjaminBrashaw — APPROVED** [review]

**Review by BenjaminBrashaw — APPROVED** [review]
Had a read through the code and everything looks okay

File: src/main/java/org/example/daos/AuthDao.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+            statement.setString(1, loginRequest.getUsername());
+
+            ResultSet resultSet = statement.executeQuery();
+
+            if (resultSet.next()) {
+                String storedPasswordHash = resultSet.getString("password");
+
+                System.out.println("hash from db: "
+                        + storedPasswordHash);
```
Comment (shaunganley) [inline src/main/java/org/example/daos/AuthDao.java]: remove

File: src/main/java/org/example/daos/AuthDao.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            } else {
+                System.err.println("User not found");
+            }
+        } catch (SQLException e) {
+            System.out.println("SQLException");
+            e.printStackTrace();
+        } catch (Exception e) {
+            System.out.println("Exception");
+            e.printStackTrace();
+        }
```
Comment (shaunganley) [inline src/main/java/org/example/daos/AuthDao.java]: don't catch generic exception

File: src/main/java/org/example/daos/AuthDao.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                    );
+                } else {
+                    System.err.println("Password verification failed");
+                }
+            } else {
+                System.err.println("User not found");
+            }
+        } catch (SQLException e) {
+            System.out.println("SQLException");
+            e.printStackTrace();
```
Comment (shaunganley) [inline src/main/java/org/example/daos/AuthDao.java]: you don't need to catch this if you're not doing anything with it

File: src/main/java/org/example/models/User.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    int loginID;
+
+    public User() {
+    }
+
+    public User(final String username, final String password,
+                final int loginID) {
+        this.username = username;
+        this.password = password;
+        this.loginID = loginID;
```
Comment (shaunganley) [inline src/main/java/org/example/models/User.java]: what is loginID and how does it differ from username?

File: src/main/java/org/example/services/AuthService.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import java.sql.SQLException;
+
+
+public class AuthService {
+
+    private final AuthDao authDao;
+    private final Key key;
+
+    private static final SecretKey SECRET_KEY = Keys.secretKeyFor(
+            SignatureAlgorithm.HS256); // Ensure proper key size and type
```
Comment (shaunganley) [inline src/main/java/org/example/services/AuthService.java]: any reason you're not using the key that you're passing in via constructor?

File: src/main/resources/db/migration/V6__create_user_table.sql [hunk 6]
```diff
@@ -0,0 +1,5 @@
+create table `User` (
+`username` varchar(100) not null,
+`password` varchar(100) not null,
+`loginID` int default 2 not null
```
Comment (shaunganley) [inline src/main/resources/db/migration/V6__create_user_table.sql]: what is this? should it be a foreign key to the UserRole table?

File: src/main/resources/db/migration/V6__create_user_table.sql [hunk 7]
```diff
@@ -0,0 +1,5 @@
+create table `User` (
+`username` varchar(100) not null,
+`password` varchar(100) not null,
```
Comment (shaunganley) [inline src/main/resources/db/migration/V6__create_user_table.sql]: do we need to store any hashes for encrypting passwords?

File: src/main/resources/db/migration/V2__adding_role_data.sql [hunk 8]
```diff
@@ -4,3 +4,26 @@ insert into Role (roleName, location, capability, band, closingDate, status) val
 insert into Role (roleName, location, capability, band, closingDate, status) values
 ('roleNameTest2', 'locationTest2', 'capabilityTest2', 'bandTest2', CURRENT_DATE(), 'closed');
 
+INSERT INTO Role (roleName, location, capability, band, closingDate, status) values
```
Comment (shaunganley) [inline src/main/resources/db/migration/V2__adding_role_data.sql]: band and capability should be a separate table with foreign keys here as we will need the ability to create, edit and delete them in the future

File: src/test/java/integration/AuthenticationIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class AuthenticationIntegrationTest {
+    private static final DropwizardAppExtension<TestConfiguration> APP =
+            new DropwizardAppExtension<>(TestApplication.class);
+
+    @Test
+    void loginShouldReturnJWTToken() {
+        Client client = APP.client();
+
+        LoginRequest loginRequest = new LoginRequest("admin", "admin");
```
Comment (shaunganley) [inline src/test/java/integration/AuthenticationIntegrationTest.java]: don't hardcode valid login credentials in code, instead pass them in as environment variables

File: src/test/java/service/AuthServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+import static org.mockito.Mockito.when;
+
+
+@ExtendWith(MockitoExtension.class)
+class AuthServiceTest {
+
+    AuthDao authDao = Mockito.mock(AuthDao.class);
+    Key key = Mockito.mock(Key.class);
+    LoginRequest loginRequest = Mockito.mock(LoginRequest.class);
+    User user = Mockito.mock(User.class);
```
Comment (shaunganley) [inline src/test/java/service/AuthServiceTest.java]: I don't think you need to mock these things. just create an object for each if you need to use them

**Review by shaunganley — CHANGES_REQUESTED** [review]
nice work, few small changes required

File: src/main/java/org/example/models/User.java (hunk 4 above)
Comment (joel-and) [reply src/main/java/org/example/models/User.java]: i have renamed this to UserRoleID, each UserRole (user, admin, applicant etc) will have a unique ID that is referenced in the user table

File: src/main/resources/db/migration/V6__create_user_table.sql (hunk 7 above)
Comment (joel-and) [reply src/main/resources/db/migration/V6__create_user_table.sql]: Passwords are stored as hashed strings in the database which is what the system compares to when a user logs in. when entering the password on front end the system hashes their password and compares it to the hashed password stored in the database

**Review by shaunganley — APPROVED** [review]

**Pull request merged**